server/src/main/java/org/elasticsearch/index/codec/TrackingPostingsInMemoryBytesCodec.java

```diff
@@ -0,0 +1,139 @@
+            }
+        };
+    }
+
+    static final class TrackingLengthFieldsConsumer extends FieldsConsumer {
+        final SegmentWriteState state;
+        final FieldsConsumer in;
+        final IntIntHashMap maxLengths;
+
+        TrackingLengthFieldsConsumer(SegmentWriteState state, FieldsConsumer in) {
+            this.state = state;
+            this.in = in;
+            this.maxLengths = new IntIntHashMap(state.fieldInfos.size());
+        }
+
+        @Override
+        public void write(Fields fields, NormsProducer norms) throws IOException {
+            in.write(new TrackingLengthFields(fields, maxLengths, state.fieldInfos), norms);
+            long totalLength = 0;
+            for (int len : maxLengths.values) {
+                totalLength += len; // minTerm
+                totalLength += len; // maxTerm
+            }
+            state.segmentInfo.putAttribute(IN_MEMORY_POSTINGS_BYTES_KEY, Long.toString(totalLength));
+        }
+
+        @Override
+        public void close() throws IOException {
+            in.close();
+        }
+    }
+
+    static final class TrackingLengthFields extends FilterLeafReader.FilterFields {
+        final IntIntHashMap maxLengths;
+        final FieldInfos fieldInfos;
+
+        TrackingLengthFields(Fields in, IntIntHashMap maxLengths, FieldInfos fieldInfos) {
+            super(in);
+            this.maxLengths = maxLengths;
+            this.fieldInfos = fieldInfos;
+        }
+
+        @Override
+        public Terms terms(String field) throws IOException {
+            Terms terms = super.terms(field);
+            if (terms == null) {
+                return terms;
+            }
+            int fieldNum = fieldInfos.fieldInfo(field).number;
+            return new TrackingLengthTerms(terms, len -> maxLengths.put(fieldNum, Math.max(maxLengths.getOrDefault(fieldNum, 0), len)));
+        }
+    }
+
+    static final class TrackingLengthTerms extends FilterLeafReader.FilterTerms {
+        final IntConsumer onFinish;
+
+        TrackingLengthTerms(Terms in, IntConsumer onFinish) {
+            super(in);
+            this.onFinish = onFinish;
+        }
+
+        @Override
+        public TermsEnum iterator() throws IOException {
+            return new TrackingLengthTermsEnum(super.iterator(), onFinish);
+        }
+    }
+
+    static final class TrackingLengthTermsEnum extends FilterLeafReader.FilterTermsEnum {
+        int maxTermLength = 0;
+        final IntConsumer onFinish;
+
+        TrackingLengthTermsEnum(TermsEnum in, IntConsumer onFinish) {
+            super(in);
+            this.onFinish = onFinish;
+        }
+
+        @Override
+        public BytesRef next() throws IOException {
+            final BytesRef term = super.next();
+            if (term != null) {
+                maxTermLength = Math.max(maxTermLength, term.length);
+            } else {
+                onFinish.accept(maxTermLength);
+            }
+            return term;
+        }
```

Comment:
Given that we need to estimate the terms that get loaded in jvm heap would the following be more accurate?

```suggestion
         int prevTermLength = 0;

        @Override
        public BytesRef next() throws IOException {
            final BytesRef term = super.next();
            if (term == null) {
                maxTermLength += prevTermLength;
                onFinish.accept(maxTermLength);
                return term;
            }
            if (maxTermLength == 0) {
                maxTermLength = term.length;
            }
            prevTermLength = term.length;
            return term;
        }
```

In the `org.apache.lucene.codecs.lucene90.blocktree.FieldReader` class, the lowest and highest lexicographically term is kept around in jvm heap. The current code just keeps track what the longest term is and report that, which doesn't map with the `minTerm` and `maxTerm` in `FieldReader`?